crates/tinymist/src/tool/preview.rs

```diff
@@ -349,61 +371,134 @@ impl PreviewState {
+                    Ok::<_, std::convert::Infallible>(res)
+                } else {
+                    // jump to /
+                    let res = hyper::Response::builder()
+                        .status(hyper::StatusCode::FOUND)
+                        .header(hyper::header::LOCATION, "/")
+                        .body(Full::<Bytes>::default())
+                        .unwrap();
+                    Ok(res)
                 }
-            }))
-        }
-    });
-    let server = hyper::Server::bind(&static_file_addr.parse().unwrap()).serve(make_service);
-    let addr = server.local_addr();
+            }
+        })
+    };
 
-    let (tx, rx) = tokio::sync::oneshot::channel();
+    let listener = tokio::net::TcpListener::bind(&static_file_addr)
+        .await
+        .unwrap();
+    let addr = listener.local_addr().unwrap();
+    log::info!("Listening preview server on http://{addr}");
+
+    let (_tx, rx) = tokio::sync::oneshot::channel();
     let (final_tx, final_rx) = tokio::sync::oneshot::channel();
-    let graceful = server.with_graceful_shutdown(async {
-        final_rx.await.ok();
-        log::info!("Static file server stop requested");
-    });
 
-    let join_handle = tokio::spawn(async move {
-        tokio::select! {
-            Err(err) = graceful => {
-                log::error!("Static file server error: {err:?}");
+    // the graceful watcher
+    let graceful = hyper_util::server::graceful::GracefulShutdown::new();
+
+    let serve_conn = move |server: &Server, graceful: &GracefulShutdown, conn| {
+        let (stream, _peer_addr) = match conn {
+            Ok(conn) => conn,
+            Err(e) => {
+                log::error!("accept error: {e}");
+                return;
+            }
+        };
+
+        let conn = server.serve_connection_with_upgrades(TokioIo::new(stream), make_service());
+        let conn = graceful.watch(conn.into_owned());
+        tokio::spawn(async move {
+            if let Err(err) = conn.await {
+                log::error!("Error serving connection: {err:?}");
+            }
+        });
+    };
+
+    let join = tokio::spawn(async move {
+        // when this signal completes, start shutdown
+        let mut signal = std::pin::pin!(final_rx);
+
+        let mut server = Server::new(hyper_util::rt::TokioExecutor::new());
+        server.http1().keep_alive(true);
+
+        loop {
+            tokio::select! {
+                conn = listener.accept() => serve_conn(&server, &graceful, conn),
+                Ok(_) = &mut signal => {
+                    log::info!("graceful shutdown signal received");
+                    break;
+                }
             }
-            _ = rx => {
-                final_tx.send(()).ok();
+        }
+
+        tokio::select! {
+            _ = graceful.shutdown() => {
+                log::info!("Gracefully shutdown!");
+            },
+            _ = tokio::time::sleep(reflexo::time::Duration::from_secs(10)) => {
```

Comment:
reflexo has its own time? interesting

Reply:
it is an alias of https://docs.rs/crate/instant/latest which will work in browser as well. Not a real bonus, but we may be able to compile typst-preview to wasm now, since all system dependencies are already taken from the preview crate.